sdk/program/src/vote/state/mod.rs

```diff
@@ -374,50 +372,64 @@ impl VoteState {
         3762 // see test_vote_state_size_of.
     }
 
-    // we retain bincode deserialize for not(target_os = "solana")
-    // because the hand-written parser does not support V0_23_5
+    /// Deserializes the input buffer into a newly allocated `VoteState`
+    ///
+    /// This function is intended as a drop-in replacement for `bincode::deserialize()`.
+    /// V0_23_5 is not supported in a BPF context, but all versions are supported on non-BPF.
     pub fn deserialize(input: &[u8]) -> Result<Self, InstructionError> {
+        let mut vote_state = Self::default();
+        Self::deserialize_into(input, &mut vote_state)?;
+        Ok(vote_state)
+    }
+
+    // this only exists for the sake of the feature gated upgrade to the new parser; do not use it
+    #[doc(hidden)]
+    #[allow(clippy::used_underscore_binding)]
+    pub fn deserialize_with_bincode(_input: &[u8]) -> Result<Self, InstructionError> {
```

Comment:
i waffled a lot on whether to try and do the feature gate and vote program/runtime parser upgrade all in this pr or split into two, but the fact that some functions that want to use VoteState::deserialize() dont have access to the featureset struct made me decide the changes are too complicated and should be done separately. this function is identical to the original `VoteState::deserialize()` and i replace calls to that with this in a few places

Reply:
That makes sense.

Rather than having this temporarily-private function though, how about just having the call-sites do `bincode::deserialize` themselves?

Reply:
ah `pub(crate)` doesnt work because multiple crates and inlining bincode would mean i need to add it to and then remove it from Cargo.tomls. i think ill just do all this in one shot in #35079 rather than create an uncomfortable intermediate state